### DISCUSSION TITLE
SkyWalking supports JDK11 HTTPCLIENT

### DISCUSSION THREAD

**Discussion by HackerRookie** [apache/skywalking#10671] [Q&A] [answered]
The skywalking-agent version is: 8.4.0
JDK version: OpenJDK version "11.0.18" 2023-01-17
If agent.is_cache_enhanced_class=true, the following exception is reported:

Exception in thread "main" java.lang.IllegalAccessError: superinterface check failed: class sun.net.www.http.HttpClient (in module java.base) cannot access class org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance (in unnamed module @0x70dd7e15) because module java.base does not read unnamed module @0x70dd7e15
at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:207)
at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:193)
at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:168)
at com.example.demo.skywalking.JDKHttpClient.main(JDKHttpClient.java:33)

**Comment by wu-sheng** [comment, marked as answer]
This is clearly documented, https://skywalking.apache.org/docs/skywalking-java/v8.15.0/en/setup/service-agent/java-agent/bootstrap-plugins/

**Reply by wu-sheng** [reply]
It looks like you are using a higher version than you thought of JDK. Anyway, this is a JDK wide usage issue.

**Comment by HackerRookie** [comment]
17+ | Add --add-exports java.base/sun.net.www=ALL-UNNAMED.  

Does the JVM add this parameter?

**Reply by wu-sheng** [reply]
Like I said, this is a JDK-level thing. SkyWalking has little to do with it.

**Comment by HackerRookie** [comment]
https://skywalking.apache.org/docs/skywalking-java/v8.15.0/en/setup/service-agent/java-agent/bootstrap-plugins/
Follow the instructions on the document to configure, still cannot be solved.

class sun.net.www.http.HttpClient (in module java.base) cannot access class org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance (in unnamed module @0x70dd7e15) because module java.base does not read unnamed module @0x70dd7e15

**Reply by wu-sheng** [reply]
I can only say, you need to open `unnamed module `. Maybe `--add-exports java.base/sun.net.www=ALL-UNNAMED` or others, which are determined by the JDK.

**Comment by HackerRookie** [comment]
There is a contradiction ：Why agent.is_cache_enhanced_class=false is normal.

**Reply by wu-sheng** [reply]
I don't know that, sorry. I never used that option.
That was introduced by @kylixs through https://github.com/apache/skywalking/pull/4858 for retransform cases.

**Reply by revolyw** [reply]
@kylixs  I have the same question, could you please help analyze it?

**Comment by HackerRookie** [comment]
Take the liberty of asking
Can you have contributors test this issue?

**Reply by wu-sheng** [reply]
This is not how SkyWalking works. All contributors are volunteers, who don't have any responsibility to maintain features as always.
Especially this feature(this config) is not widely used, I can't tell who will do this.

**Comment by HackerRookie** [comment]
#4858 implemented
org.apache.skywalking.apm.agent.core.plugin.bytebuddy.CacheableTransformerDecorator does not support module 9.

**Comment by kylixs** [comment]
I am on it.

**Comment by kylixs** [comment]
> The skywalking-agent version is: 8.4.0 JDK version: OpenJDK version "11.0.18" 2023-01-17 If agent.is_cache_enhanced_class=true, the following exception is reported:
> 
> Exception in thread "main" java.lang.IllegalAccessError: superinterface check failed: class sun.net.www.http.HttpClient (in module java.base) cannot access class org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance (in unnamed module @0x70dd7e15) because module java.base does not read unnamed module @0x70dd7e15 at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:207) at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187) at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081) at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:193) at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:168) at com.example.demo.skywalking.JDKHttpClient.main(JDKHttpClient.java:33)

@HackerRookie How to reproduce the problem? Can you submit your demo to github ?
I just test this use case is no problem at https://github.com/SkyAPMTest/retransform-conflict-demo

**Comment by HackerRookie** [comment]
Turn on first agent.is_cache_enhanced_class=true
Then use JDK11. Implement if under request
@kylixs 
public static void main(String[] args) {

        InputStreamReader isr = null;

        URL url = null;
        try {
            url = new URL("https://www.baidu.com");
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        }
        HttpURLConnection conn = null;
        try {
            conn = (HttpURLConnection)url.openConnection();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        conn.setConnectTimeout(10000);
        conn.setReadTimeout(60000);
        try {
            conn.connect();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        try {
            isr = new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        try {
            if (conn.getResponseCode() == 200) {
                String var5 = readToString(isr);
               System.out.println(var5);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        conn.setConnectTimeout(10000);
        conn.setReadTimeout(60000);
        try {
            conn.connect();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

**Reply by HackerRookie** [reply]
For JDK-HTTPCLIENT plugins @kylixs

**Reply by kylixs** [reply]
Please look at this: https://github.com/apache/skywalking-java/pull/521
